src/lexer/regexFactory.ts

```diff
@@ -0,0 +1,145 @@
+
+/**
+ * Builds a RegExp for matching parenthesis patterns, escaping them with `escapeParen`
+ * @param {string[]} parens - list of strings that denote parenthesis patterns
+ */
+export const parenthesis = (parens: string[]): RegExp =>
+  patternToRegex(parens.map(escapeParen).join('|'));
+
+/**
+ * Builds a RegExp containing all operators for a SQL dialect
+ * @param {string} monadOperators - concatenated string of all 1-length operators
+ * @param {string[]} polyadOperators - list of strings of all >1-length operators
+ */
+export const operator = (monadOperators: string, polyadOperators: string[]) =>
+  patternToRegex(
+    `${sortByLengthDesc(polyadOperators).map(escapeRegExp).join('|')}|` +
+      `[${monadOperators.split('').map(escapeRegExp).join('')}]`
+  );
+
+// Negative lookahead to avoid matching a keyword that's actually part of identifier,
+// which can happen when identifier allows word-boundary characters inside it.
+//
+// For example "SELECT$ME" should be tokenized as:
+// - ["SELECT$ME"] when $ is allowed inside identifiers
+// - ["SELECT", "$", "ME"] when $ can't be part of identifiers.
+const rejectIdentCharsPattern = ({ rest, dashes }: IdentChars): string =>
+  rest || dashes ? `(?![${rest || ''}${dashes ? '-' : ''}])` : '';
+
+/**
+ * Builds a RegExp for all Reserved Keywords in a SQL dialect
+ */
+export const reservedWord = (reservedKeywords: string[], identChars: IdentChars = {}): RegExp => {
+  if (reservedKeywords.length === 0) {
+    return /^\b$/u;
+  }
+
+  const avoidIdentChars = rejectIdentCharsPattern(identChars);
+
+  const reservedKeywordsPattern = sortByLengthDesc(reservedKeywords)
+    .map(toCaseInsensitivePattern)
+    .join('|')
+    .replace(/ /gu, '\\s+');
+
+  return new RegExp(`(?:${reservedKeywordsPattern})${avoidIdentChars}\\b`, 'iu');
+};
+
+/**
+ * Builds a RegExp for parameter placeholder patterns
+ * @param {string[]} paramTypes - list of strings that denote placeholder types
+ * @param {string} pattern - string that denotes placeholder pattern
+ */
+export const parameter = (paramTypes: string[], pattern: string): RegExp | undefined => {
+  if (!paramTypes.length) {
+    return undefined;
+  }
+  const typesRegex = paramTypes.map(escapeRegExp).join('|');
+
+  return patternToRegex(`(?:${typesRegex})(?:${pattern})`);
+};
+
+// This enables the following quote styles:
+// 1. backtick quoted using `` to escape
+// 2. square bracket quoted (SQL Server) using ]] to escape
+// 3. double quoted using "" or \" to escape
+// 4. single quoted using '' or \' to escape
+// 5. PostgreSQL dollar-quoted
+// 6. BigQuery '''triple-quoted'''
+// 7. BigQuery """triple-quoted"""
+// 8. Hive and Spark variables: ${name}
+export const quotePatterns = {
+  '``': '(?:`[^`]*(?:$|`))+',
+  '[]': '(?:\\[[^\\]]*(?:$|\\]))(?:\\][^\\]]*(?:$|\\]))*',
+  '""': '(?:"[^"\\\\]*(?:\\\\.[^"\\\\]*)*(?:"|$))+',
+  "''": "(?:'[^'\\\\]*(?:\\\\.[^'\\\\]*)*(?:'|$))+",
+  // '$$': '(?<tag>\\$\\w*\\$)[\\s\\S]*?(?:\\k<tag>|$)',
```

Comment:
For me, this is a serious blocker. We can't use a lexer which doesn't allow us to parse languages which we're targeting. We either need a workaround for this or we can't use Moo.

It's not only PostgreSQL either. There's a similar string syntax in Oracle SQL too, for which we haven't yet added support (`q'x..x'` where `x` can be any character).